facade/service.go

```diff
@@ -1368,21 +1411,139 @@ func (f *Facade) ScheduleServiceBatch(ctx datastore.Context, svcs []*service.Ser
 	defer ctx.Metrics().Stop(ctx.Metrics().Start("Facade.updateDesiredState"))
-	switch desiredState {
-	case service.SVCRestart:
-		// shutdown all service instances
-		if err := f.zzk.StopServiceInstances(ctx, svc.PoolID, svc.ID); err != nil {
+	logger := plog.WithFields(log.Fields{
+		"serviceid":    svc.ID,
+		"servicename":  svc.Name,
+		"desiredstate": desiredState,
+	})
+	if desiredState == service.SVCRestart && svc.DesiredState == int(service.SVCStop) {
+		// If we are already stopped or stopping, just start it back up
+		desiredState = service.SVCRun
+	}
+
+	if desiredState == service.SVCRestart {
+		if err := f.rollingRestart(ctx, svc, f.rollingRestartTimeout); err != nil {
+			logger.WithError(err).Debug("Could not perform rolling restart for service")
 			return err
 		}
-		svc.DesiredState = int(service.SVCRun)
-	default:
+	} else {
 		svc.DesiredState = int(desiredState)
+		// write the service into the database
+		if err := f.serviceStore.UpdateDesiredState(ctx, svc.ID, svc.DesiredState); err != nil {
+			logger.WithError(err).Debug("Could not update desired state")
+			return err
+		}
 	}
 
-	// write the service into the database
-	if err := f.serviceStore.UpdateDesiredState(ctx, svc.ID, svc.DesiredState); err != nil {
-		glog.Errorf("Facade.updateDesiredState: Could not create service %s (%s): %s", svc.Name, svc.ID, err)
-		return err
+	return nil
+}
+
+// rollingRestart restarts a service one instance at a time and waits for it to reach pass health checks.
+func (f *Facade) rollingRestart(ctx datastore.Context, svc *service.Service, timeout time.Duration) error {
+	logger := plog.WithFields(log.Fields{
+		"serviceid":   svc.ID,
+		"servicename": svc.Name,
+	})
+
+	// Build the service health object to use for getting instance health
+	svch := service.BuildServiceHealth(*svc)
+
+	for instanceID := 0; instanceID < svc.Instances; instanceID++ {
+		ilogger := logger.WithField("instance", instanceID)
+		ilogger.Debug("Restarting instance")
+		// Set up the timeout
+		cancel := make(chan struct{})
+		cancelWait := make(chan struct{})
+		timer := time.NewTimer(timeout)
+		go func() {
+			defer timer.Stop()
+
+			select {
+			case <-timer.C:
+				ilogger.Warn("Timeout waiting for instance to restart")
+				close(cancelWait)
+			case <-cancel:
+				close(cancelWait)
+			}
+		}()
+
+		// We need to start the wait in a goroutine before calling restart, so we can be sure to catch the first
+		// transition from running to stopped
+		errC := make(chan error)
+		go func() {
+			// First wait for the stop
+			err := f.zzk.WaitInstance(ctx, svc, instanceID, service.SVCStop, cancelWait)
+			if err != nil {
+				errC <- err
+				return
+			}
+
+			// If we got cancelled, bail:
+			select {
+			case <-cancelWait:
+				errC <- nil
+				return
+			default:
+			}
+
+			// Then for the start
+			errC <- f.zzk.WaitInstance(ctx, svc, instanceID, service.SVCRun, cancelWait)
```

Comment:
Yeah, that could happen.  We should probably implement a cancel mechanism that could be initiated by the service state manager when a new state change request comes in.  I think we will still have to deal with this even when we switch to monitoring the ContainerID instead of WaitInstance, correct?